ZukoAzula2016/src/org/team1540/zukoazula/Portcullis.java

```diff
@@ -20,43 +25,72 @@
+
+    private static final ArbitratedFloat leftInput = grabBehaviors.addFloat();
+    private static final ArbitratedFloat rightInput = grabBehaviors.addFloat();
+
+    private static final BooleanCell needsToCalibrate = new BooleanCell(true);
 
     private static final FloatInput pidP = ZukoAzula.mainTuning.getFloat("Portcullis PID:P", 0.001f);
     private static final FloatInput pidI = ZukoAzula.mainTuning.getFloat("Portcullis PID:I", 0.0f);
     private static final FloatInput pidD = ZukoAzula.mainTuning.getFloat("Portcullis PID:D", 0.0f);
 
     private static final FloatInput autolevelSpeed = ZukoAzula.mainTuning.getFloat("Portcullis Auto-level Speed", 0.2f);
-    private static final FloatCell maximumSpeed = ZukoAzula.mainTuning.getFloat("Portcullis Maximum Speed", 0.4f);
+
+    private static final FloatCell autonomousVelocity = new FloatCell();
+    private static FloatInput position;
 
     public static void setup() throws ExtendedMotorFailureException {
+        position = leftEncoder.normalize(ZukoAzula.mainTuning.getFloat("Portcullis Distance to Low Position", 2470), ZukoAzula.mainTuning.getFloat("Portcullis Distance to High Position", -100));
 
         PIDController levelPID = new PIDController(leftEncoder, leftEncoder.plus(rightEncoder.negated()).dividedBy(2), pidP, pidI, pidD);
         levelPID.updateWhen(FRC.constantPeriodic);
         levelPID.setOutputBounds(autolevelSpeed);
-        
-        leftEncoder.eventSet(rightEncoder.negated()).on(FRC.startAuto.or(FRC.startTele));
-        
-        FloatInput leftOut = control.multipliedBy(maximumSpeed).minus(control.outsideRange(-0.1f, 0.1f).toFloat(levelPID, 0.0f));
-        FloatInput rightOut = control.multipliedBy(maximumSpeed).plus(control.outsideRange(-0.1f, 0.1f).toFloat(levelPID, 0.0f));
 
-        leftInput.attach(ZukoAzula.teleop, leftOut);
-        rightInput.attach(ZukoAzula.teleop, rightOut);
-        leftInput.attach(ZukoAzula.pit, leftOut);
-        rightInput.attach(ZukoAzula.pit, rightOut);
+        BooleanInput calibrating = needsToCalibrate.and(FRC.inTeleopMode().or(FRC.inAutonomousMode()));
+        EventOutput resetEncoders = leftEncoder.eventSet(0).combine(rightEncoder.eventSet(0)).combine(needsToCalibrate.eventSet(false));
+        FloatInput stalling = ZukoAzula.mainTuning.getFloat("Portcullis Stalling Current Threshold", 1.75f);
+        BooleanInput bothStalling = leftOutputCurrent.atLeast(stalling).and(rightOutputCurrent.atLeast(stalling));
+        resetEncoders.on(bothStalling.onPress().and(calibrating));
+
+        Behavior teleop = grabBehaviors.addBehavior("teleop", FRC.inTeleopMode());
+        leftInput.attach(teleop, targetVelocity.minus(targetVelocity.inRange(-0.1f, 0.1f).toFloat(0.0f, levelPID)));
+        rightInput.attach(teleop, targetVelocity.plus(targetVelocity.inRange(-0.1f, 0.1f).toFloat(0.0f, levelPID)));
+        Behavior autonomous = grabBehaviors.addBehavior("autonomous", FRC.inAutonomousMode());
```

Comment:
Again, considering that these are indeed global modes, I agree with @Ja-ake that it should be consolidated.